### ISSUE TITLE
Exclude Deppbot branches from Travis

### ISSUE THREAD

**Issue by lowjoel** [deppbot/feedback#1]
If you've set up Travis, then this extra config in .travis.yml will ensure that the Deppbot branches aren't built twice (and saves strain on Travis' infrastructure)

``` yml
branches:
  except:
    - /^bundle-update-[0-9]{4}-[0-9]{2}-[0-9]{2}-[0-9]{6}+$/
```

Maybe put this somewhere? If it is already in the docs somewhere I missed it...

**Comment by JuanitoFatas** [comment]
Hi @lowjoel!

Thanks for your feedback.

deppbot only send you Pull Request, deppbot did not run your test suite, so your CI should run against @deppbot's Pull Request.

**Comment by lowjoel** [comment]
Yes. But what happens is that Travis will build ALL branches in a repo by default. So when deppbot creates the bundle update branch, Travis will build that branch.

Then when deppbot creates the PR, Travis will build the PR also.

In total, Travis builds that branch twice. See Coursemology/coursemology2#760, look at the number of status checks. Travis appears twice.

**Comment by JuanitoFatas** [comment]
> In total, Travis builds that branch twice. See Coursemology/coursemology2#760, look at the number of status checks. Travis appears twice.

Oh that's two different things to be tested, see https://github.com/travis-ci/travis-ci/issues/3241#issuecomment-77122258.

Does above comment help? Thanks!

**Comment by lowjoel** [comment]
Exactly. That's why I'm giving this additional snippet so that the PR is only built once -- this could be part of the set up instructions so that people won't get into this situation.

**Comment by winston** [comment]
@lowjoel I don't think this is the correct advice though, and the two builds should be ran.

Although it may seem like `travis-ci/pr` and `travis-ci/push` are the same, but they are actually not. 

We have seen plenty of cases like https://github.com/jollygoodcode/dasherize/pull/41, where users work on the branch for various reasons (make tests pass especially). Hence it's important to actually run the the branch (`travis-ci/push`).

`travis-ci/pr` is testing the PR's merge commit to base branch, and the base branch can be different from when the bundle update was issued.

So it's better to let travis do its job and run both builds and which is what travis ci would want to do (that's in fact it's USP against its competition).

Hope this clarifies.

**Comment by lowjoel** [comment]
Hmm, even after disabling the push build, if there are any updates to the branch, the `pr` build will still run. That issue says that one tests the merge commit, the other tests the tip of the branch.

Is there another example of when the `pr` and `push` builds will build differently? The example of pushing to the branch will still be accounted for because it's really no different from a PR from a repo fork, where only the `pr` build runs. That shouldn't be a problem because we're really concerned about the mergeability of the two branches and whether tests pass on the new merged branch (...right?)

I think the only case where it the result would be different is when the HEAD branch moves on from the base commit of the deppbot branch (like you mentioned), but in those situations the deppbot branch won't have its build triggered automatically, either. So building both sets of changes would be superfluous (IMO, anyway)

Did I miss anything?

Even if Travis really wants to run all those tests, one set of tests take ~10 minutes to run on my end, and it really tests patience to have twice as many builds blocking other builds/PRs going on.

At the end of the day, I'd thought I'd just give a tip that took me a while to figure out to other people who might have this same problem (I was browsing Travis docs), not a problem if you all think that building both is more correct and is recommended.

**Comment by lowjoel** [comment]
@winston I was thinking about this further, and I think it's because I'm coming from a strict PR workflow I'm suggesting this. Maybe what you're saying is that if people directly push to a repository, this won't work so well?

I have pushed straight to a Deppbot branch before, and the status checks are triggered again, so I think that shouldn't be a problem. I could dogfood this idea on my own repo and see how it goes, too.

**Comment by winston** [comment]
> At the end of the day, I'd thought I'd just give a tip that took me a while to figure out to other people who might have this same problem

We definitely welcome the discussion! :smile: I think this is more a travis than a deppbot problem though. 
And to travis, it's more of a feature than a problem.  

Maybe it's better to ask on travis what's the rationale for having the 2 builds?

To us, it doesn't seem right to suggest how our users should run travis because our aim for deppbot is to try and mimic what a human on the team would do: 
1) run bundle update 
2) open a pr. 

So it's strange if we suggest that deppbot branches are disabled for the PR merge or push commit run. Does that mean other branches by other members of the team should do that do? Otherwise, why only disable for deppbot branches? (and hence it's more of understanding travis's rationale).

>  I could dogfood this idea on my own repo and see how it goes, too.

We would definitely love to hear about your findings!

**Comment by lowjoel** [comment]
Hmm, if the team uses PRs exclusively (like we do), then it would be like Deppbot forking a repo, push the changes, and open a PR. The PR won't be in the same repository as every other bit of code, and so only the `pr` task will run on Travis, hence it's a bit weird to need to run all specs twice.

I've pushed this config up to all my repos, I'll see how it plays out over the next few days.

**Comment by choonkeat** [comment]
Great tip @lowjoel albeit not deppbot specific :-) I think it makes sense for PR workflow - and would be a nice Travis feature/option

**Comment by lowjoel** [comment]
So I've run this set up for a while, on my PR-only workflow. Excluding the branches is positive. Even when we need to push to the Deppbot branch, Travis still builds the PR (no surprises). Rebasing to squash commits on the branch works also.

So there isn't any detriment in enabling this for us, anyway. Those who directly commit to their repos might also benefit, but since we don't have that workflow, I can't say for sure.

This could be a tip of sorts for people who use PR workflows? I'm not sure.

**Comment by winston** [comment]
I looked at the PRs on https://github.com/Coursemology/coursemology2 and realised that they actually came from another repo https://github.com/Coursemology-collab, and so the branches are actually on collab and not on the main coursemology2 repo. 

Hence for your PRs, you are only getting 1 CI run vs deppbot's PR which has two runs, because deppbot creates the branch within the current repo (1 CI run for the branch) and then creates the PR (another CI run).

So in a way, because of how you manage your PR (coming from forks), this results in the "correct" travis behavior where there's only 1 run (the PR run). But if you look at this PR https://github.com/rubysg/rubysg-reboot/pull/18, where I didn't fork and I created the branch on the main repo and then issued a PR, there are two CI runs too (1 for branch and 1 for PR). This means that if, one day, you also create a branch on coursemology2 and issue a PR from that branch, the behavior of 2 CI runs for that PR will happen again as well - which you would feel is unnecessary.

In short, the proposed config really only works specifically for your case, and hence not really deppbot related, although deppbot is the oddball here because deppbot doesn't fork. :)

Hope this explains. I am going to close this now. Thanks @lowjoel !
